lib/collection/src/common/stoppable_task_async.rs

```diff
@@ -33,38 +32,35 @@ impl<T: Clone> StoppableAsyncTaskHandle<T> {
     }
 }
 
-pub fn spawn_async_stoppable<F, T>(f: F) -> StoppableAsyncTaskHandle<T::Output>
+pub fn spawn_async_cancellable<F, T>(f: F) -> CancellableAsyncTaskHandle<T::Output>
 where
-    F: FnOnce(Arc<AtomicBool>) -> T,
+    F: FnOnce(CancellationToken) -> T,
     F: Send + 'static,
     T: Future + Send + 'static,
     T::Output: Clone + Send + 'static,
 {
+    let cancelled = CancellationToken::new();
     let finished = Arc::new(AtomicBool::new(false));
-    let finished_c = finished.clone();
-
-    let stopped = Arc::new(AtomicBool::new(false));
-    // We are OK if original value is destroyed with the thread
-    // Weak reference is sufficient
-    let stopped_w = Arc::downgrade(&stopped);
-
     let result_holder = Arc::new(Mutex::new(None));
-    let result_holder_c = result_holder.clone();
-
-    StoppableAsyncTaskHandle {
-        join_handle: tokio::task::spawn(async move {
-            let res = f(stopped).await;
-            let mut result_holder_w = result_holder_c.lock();
-            result_holder_w.replace(res.clone());
-
-            // We use `Release` ordering to ensure that `f` won't be moved after the `store`
-            // by the compiler
-            finished.store(true, Ordering::Release);
-            res
+
+    CancellableAsyncTaskHandle {
+        join_handle: tokio::task::spawn({
+            let (cancel, finished, result_holder) =
+                (cancelled.clone(), finished.clone(), result_holder.clone());
+            async move {
+                let res = f(cancel).await;
+                let mut result_holder_w = result_holder.lock();
+                result_holder_w.replace(res.clone());
+
+                // We use `Release` ordering to ensure that `f` won't be moved after the `store`
+                // by the compiler
+                finished.store(true, Ordering::Release);
+                res
+            }
```

Comment:
TBH, I don't understand, why we bother with `finished` and `result_holder` at this point. `JoinHandle` already provides both of these.

- we can use [`JoinHandle::is_finished`] instead of `finished`
- and we can use [`JoinHandle.await`] instead of `result_holder`
  - it's trivial to do a `try_ready` check with a `Future`, if we don't want to do a full `.await`

[`JoinHandle::is_finished`]: https://docs.rs/tokio/latest/tokio/task/struct.JoinHandle.html#method.is_finished
[`JoinHandle.await`]: https://docs.rs/tokio/latest/tokio/task/struct.JoinHandle.html#impl-Future-for-JoinHandle%3CT%3E

Reply:
Thank you for diving into this. I'm not sure why this was implemented this way, maybe we have a good reason for it. `git blame` shows that Andrey knows more about this.

In this PR I simply left it the way we had but ported the atomic boolean to the cancellation token.

Let's dive into t his in another PR.

Reply:
Quick note, I don't think our `finish` is the same as `is_finished`. Our finish will not be `true` if we abort the task, while `is_finished()` will be `true` eventually.